sbi/diagnostics/lc2st.py

```diff
@@ -0,0 +1,730 @@
+
+
+class LC2ST_NF(LC2ST):
+    def __init__(
+        self,
+        thetas: Tensor,
+        xs: Tensor,
+        posterior_samples: Tensor,
+        flow_inverse_transform: Callable[[Tensor, Tensor], Tensor],
+        flow_base_dist: torch.distributions.Distribution,
+        n_eval: int = 10_000,
+        trained_clfs_null: Optional[Dict[str, List[Any]]] = None,
+        **kwargs: Any,
+    ) -> None:
+        """
+        L-C2ST for normalizing flows.
+
+        LC2ST_NF is a subclass of LC2ST that performs the test in the space of the
+        base distribution of a normalizing flow. It uses the inverse transform of the
+        normalizing flow to map the samples from the prior and the posterior to the
+        base distribution space. Important features are:
+
+            - the null distribution is the base distribution of the normalizing flow.
+            - no `P_eval` is passed to the evaluation functions (e.g. `_scores`),
+                as the base distribution is known, samples are drawn at initialization.
+            - no permutation method is used, as the null distribution is known,
+                samples are drawn during `train_null`.
+            - the classifiers can be pre-trained under the null and `trained_clfs_null`
+                passed as an argument at initialization. They do not depend on the
+                observed data (i.e. `posterior_samples` and `xs`).
+
+        Args:
+            thetas: Samples from the prior, of shape (sample_size, dim).
+            xs: Corresponding simulated data, of shape (sample_size, dim_x).
+            posterior_samples: Samples from the estiamted posterior,
+                of shape (sample_size, dim).
+            flow_inverse_transform: Inverse transform of the normalizing flow.
+                Takes thetas and xs as input and returns noise.
+            flow_base_dist: Base distribution of the normalizing flow.
+            n_eval: Number of samples to evaluate the L-C2ST.
+            trained_clfs_null: Pre-trained classifiers under the null.
+            kwargs: Additional arguments for the LC2ST class.
+        """
+        # Aplly the inverse transform to the thetas and the posterior samples
+        self.flow_inverse_transform = flow_inverse_transform
+        inverse_thetas = flow_inverse_transform(thetas, xs).detach()
+        inverse_posterior_samples = flow_inverse_transform(
+            posterior_samples, xs
+        ).detach()
+
+        # Initialize the LC2ST class with the inverse transformed samples
+        super().__init__(inverse_thetas, xs, inverse_posterior_samples, **kwargs)
+
+        # Set the parameters for the null hypothesis testing
+        self.null_distribution = flow_base_dist
+        self.permutation = False
+        self.trained_clfs_null = trained_clfs_null
+
+        # Draw samples from the base distribution for evaluation
+        self.P_eval = flow_base_dist.sample(torch.Size([n_eval]))
+
+    def _scores(
```

Comment:
maybe I am missing something, but to me it seems all the class methods below are just calling `super.method(...)`. 
Unless there is something flow-related that needs to be changed, you could just delete them here, no?

Reply:
They use the samples from the base distribution for `P_eval` (or now `theta_o`). So basically you are right, the methods call the parent method but with specified `self.P`. This is just to avoid misusing the NF-method and passing wrong `P_eval`.

Reply:
No, there is no other scenario, because the base distribution from which we get the new `theta_o` is independent of `x_o`. But I can make it the user's responsibility. You tell me!

Reply:
I am not sure I understand. In the parent class, `theta_o` are samples from the posterior given `x_o`, but here this changes so that `theta_o` are samples from the underlying base distribution? I think this would need a bit more comments and mentioning in the docstrings. At the moment, the class docstring says that samples from the prior and the posterior (i.e., `theta_o`) are inverse transformed and evaluated under the `base`, but then `theta_o` is initialized from the `base` in `__init__`?

Reply:
Can you please help here @JuliaLinhart? I am missing something, or was this resolved?

Reply:
Okay so basically a sample from the flow is obtained by pushing a sample from the base distribution through the forward flow transformation: $\theta = T_\phi(z; x_o) \sim q_\phi(\theta \mid x_o)$. So applying the inverse transformation to this sample, recovers the sample from the base distribution. In the end, the `LC2ST_NF` method compares samples from the base distribution $z_{c=0} \sim \mathcal{N}(0,1)$ (class $c=0$) and samples from the true posterior $\theta_o \sim p(\theta \mid x_o)$ inversely transformed by the flow (class $c=1$):
$$z_{c=1} = T_\phi^{-1}(\theta_o; x_o) \sim p(T_\phi^{-1}(\theta; x_o) \mid x_o)$$

Reply:
Does this clarify things? You can also look at the tutorial or the paper itself, where this `-NF` version is explained more in detail. 

Let me know if I should change the code and / or add a description.

Reply:
ok, make sense now. thanks for clarifying!